runtime/lua/vim/cache.lua

```diff
@@ -0,0 +1,481 @@
+end
+
+--- `loadfile` using the cache
+---@param filename? string
+---@param mode? "b"|"t"|"bt"
+---@param env? table
+---@return function?, string?  error_message
+---@private
+-- luacheck: ignore 312
+function Cache.loadfile(filename, mode, env)
+  local start = uv.hrtime()
+  filename = Cache.normalize(filename)
+  mode = nil -- ignore mode, since we byte-compile the lua source files
+  local chunk, err = M.load(filename, { mode = mode, env = env })
+  M.track('loadfile', start)
+  return chunk, err
+end
+
+--- Checks whether two cache hashes are the same based on:
+--- * file size
+--- * mtime in seconds
+--- * mtime in nanoseconds
+---@param h1 CacheHash
+---@param h2 CacheHash
+---@private
+function Cache.eq(h1, h2)
+  return h1
+    and h2
+    and h1.size == h2.size
+    and h1.mtime.sec == h2.mtime.sec
+    and h1.mtime.nsec == h2.mtime.nsec
+end
+
+--- Loads the given module path using the cache
+---@param modpath string
+---@param opts? {hash?: CacheHash, mode?: "b"|"t"|"bt", env?:table} (table|nil) Options for loading the module:
+---    - hash: (table) the hash of the file to load if it is already known. (defaults to `vim.loop.fs_stat({modpath})`)
+---    - mode: (string) the mode to load the module with. "b"|"t"|"bt" (defaults to `nil`)
+---    - env: (table) the environment to load the module in. (defaults to `nil`)
+---@see |luaL_loadfile()|
+---@return function?, string? error_message
+function M.load(modpath, opts)
+  local start = uv.hrtime()
+
+  opts = opts or {}
+  local hash = opts.hash or uv.fs_stat(modpath)
+  ---@type function?, string?
+  local chunk, err
+
+  if not hash then
+    -- trigger correct error
+    chunk, err = Cache._loadfile(modpath, opts.mode, opts.env)
+    M.track('load', start)
+    return chunk, err
+  end
+
+  local entry = Cache.read(modpath)
+  if entry and Cache.eq(entry.hash, hash) then
+    -- found in cache and up to date
+    -- selene: allow(incorrect_standard_library_use)
+    chunk, err = load(entry.chunk --[[@as string]], '@' .. modpath, opts.mode, opts.env)
+    if not (err and err:find('cannot load incompatible bytecode', 1, true)) then
+      M.track('load', start)
+      return chunk, err
+    end
+  end
+  entry = { hash = hash, modpath = modpath }
+
+  chunk, err = Cache._loadfile(modpath, opts.mode, opts.env)
+  if chunk then
+    entry.chunk = string.dump(chunk)
+    Cache.write(modpath, entry)
+  end
+  M.track('load', start)
+  return chunk, err
+end
+
+--- Finds the module path for the given module name
+---@param modname string
+---@param opts? CacheFindOpts (table|nil) Options for finding a module:
+---    - rtp: (boolean) Search for modname in the runtime path (defaults to `true`)
+---    - patterns: (string[]) Paterns to use (defaults to `{"/init.lua", ".lua"}`)
+---    - paths: (string[]) Extra paths to search for modname (defaults to `{}`)
+---@return string? modpath, CacheHash? hash (string|nil) modpath for the module
+function M.find(modname, opts)
```

Comment:
Does this also check `rtp` is there is a cache miss?

Reply:
It checks the rtp by default, unless `rtp=false`. Is that what you mean?

I use it in lazy to find modules from plugins that are not yet loaded. For that I pass `rtp=false` and `paths=paths_of_unloaded_plugins`